### PR TITLE
network: wait for IPv6 MTU being synced to link MTU

### PR DESCRIPTION
**Pull request by yuwata**
The kernel resets the IPv6 MTU of an interface when its link MTU is changed.
But it seems the operation is asynchronous, and even when we detect that
the link MTU is changed, the IPv6 MTU may not be reset yet.
```
[ 2257.067613] systemd-networkd[447122]: veth99: MTU is changed: 1500 →1600 (min: 68, max: 65535)
[ 2257.067641] systemd-networkd[447122]: Setting '/proc/sys/net/ipv6/conf/veth99/mtu' to '1410'
[ 2257.067711] systemd-networkd[447122]: No change in value '1410', suppressing write
```
As you can see, even if the link MTU is changed to 1600, the IPv6 MTU is
unchanged (in this case, still 1410).

Hopefully fixes the failure found in
https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-vagrant-archlinux-sanitizers/17882/artifact/systemd-centos-ci/index.html

### CODE REVIEW

File: src/basic/sysctl-util.c [hunk 1]
```diff
@@ -193,3 +194,16 @@ int sysctl_read_ip_property(int af, const char *ifname, const char *property, ch
...
+        int r;
+
+        assert(ret);
+
+        r = sysctl_read_ip_property(af, ifname, property, &s);
+        if (r < 0)
+                return r;
+
+        return safe_atou32(s, ret);
+}
```
Comment (poettering) [inline src/basic/sysctl-util.c:222]: maybe, add:

```c
#define sysctl_read_ip_property_uint(af,ifname,property,ret) sysctl_read_ip_property_uint32(af,ifname,property,ret)
```

and then use that in routes_max() in src/network/networkd-route-util.c.

File: src/network/networkd-sysctl.c [hunk 2]
```diff
@@ -511,6 +512,14 @@ int link_set_ipv6_mtu(Link *link, int log_level) {
         if (!link_is_configured_for_family(link, AF_INET6))
                 return 0;
 
+        if (!IN_SET(link->state, LINK_STATE_CONFIGURING, LINK_STATE_CONFIGURED))
+                return 0;
+
+        if (sd_event_source_get_enabled(link->ipv6_mtu_wait_synced_event_source, NULL) > 0) {
```
Comment (poettering) [inline src/network/networkd-sysctl.c]: doesn't matter, but maybe add `/* ret= */`

File: src/network/networkd-sysctl.c [hunk 3]
```diff
@@ -534,6 +543,83 @@ int link_set_ipv6_mtu(Link *link, int log_level) {
...
+                return 1; /* done */
+        }
+
+        /* Check if IPv6 MTU is synced. */
+        r = sysctl_read_ip_property_uint32(AF_INET6, link->ifname, "mtu", &current_mtu);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "Failed to read IPv6 MTU: %m");
+
+        if (current_mtu == link->mtu) {
+                 /* Already synced. Update IPv6 MTU now. */
```
Comment (poettering) [inline src/network/networkd-sysctl.c]: one space too much indentation here

File: src/network/networkd-sysctl.c [hunk 4]
```diff
@@ -534,6 +543,83 @@ int link_set_ipv6_mtu(Link *link, int log_level) {
...
+        }
+
+        /* If not, set up a timer event source. */
+        r = event_reset_time_relative(
+                        link->manager->event, &link->ipv6_mtu_wait_synced_event_source,
+                        CLOCK_BOOTTIME, 100 * USEC_PER_MSEC, 0,
+                        ipv6_mtu_wait_synced_handler, link,
+                        /* priority = */ 0, "ipv6-mtu-wait-synced", /* force = */ true);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "Failed to configure timer event source for waiting for IPv6 MTU being synced: %m");
```
Comment (poettering) [inline src/network/networkd-sysctl.c:587]: hmm, uh, a timer? doesn't the kernel send another netlink msg when it is done updating the mtu?

Comment (yuwata) [reply src/network/networkd-sysctl.c:587]: It seems, at least judging from the log. Though, I've not checked the kernel code yet.

**Pull request merged**